sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,814 @@
+  multiple of the allocation chunk size.
+* When an asynchronous free command is enqueued, the USM pointer to be freed is
+  immediately no longer valid. The asynchronous free command may execute any
+  time between the command being enqueued and the command being synchronized
+  with. Once the asynchronous free command is enqueued memory allocated to that
+  USM pointer is now free for the memory pool to re-allocate to another USM
+  pointer either once the command has completed, or if an asynchronous command
+  is dependent on this asynchronous free command.
+* The memory pool will not release memory in the pool that is currently
+  allocated to a USM pointer, however, it may release any free memory in the
+  pool. When the memory pool releases memory from the pool, this is done so in
+  multiples of the allocation chunk size.
+* The memory pool should try to maintain an amount of memory in the pool
+  set by the threshold size in order to better facilitate re-use of the memory
+  between subsequent asynchronous malloc commends, however, it is not required
+  to.
+* The amount of total and free memory reported by the memory pool is updated
+  immediately after returning from calls to enqueue asynchronous malloc and free
+  commands, even if the command does not execute immediately.
+* If the memory is not constructed from a user provided memory allocation, when
+  the threshold size of the memory pool is set, the value specified will round
+  down to the nearest multiple of the allocation chunk size and the memory pool
+  will immediately start using this new size.
+* When the maximum size of the memory pool is set, the value specified will
+  round up to the nearest multiple of the allocation chunk size.
+
+Memory pools are intended to be used for both in-order and out-of-order SYCL
+queues.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+class memory_pool {
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, Properties props = {});
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(queue q, usm::alloc kind, Properties props = {});
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, void *ptr, size_t size, Properties props = {});
+
+  ~memory_pool();
+
+  context get_context() const;
+
+  device get_device() const;
+
+  usm::alloc get_alloc_kind() const;
+
+  size_t get_allocation_chunk_size() const;
+
+  size_t get_max_size() const;
+
+  size_t get_threshold() const;
+
+  size_t get_current_size() const;
+
+  size_t get_free_size() const;
+
+  void set_new_threshold(size_t newThreshold);
+
+}; // memory_pool
+
+}  // ext::oneapi::experimental
+----
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(context ctx, Properties props = {});
```

Comment:
I made this comment also in an earlier draft.  I think this constructor will be confusing because it is not obvious that it allocates "host" USM.  I think it might be better to add a `kind` parameter, and then also throw an exception if `kind` is not `usm::alloc::host`.

This also allows you to use this constructor in the future for `usm::alloc::shared` when there is no associated device.  Level Zero has the ability today to create "shared" USM memory that is not associated with a device, so it's possible we may add that ability to SYCL someday.